import os
import math
import logging
import anndata
import numpy as np
import pandas as pd
import scanpy as sc
import scipy
import random
import tensorflow as tf

import matplotlib.pyplot as plt
plt.rcParams.update({'font.size': 18})

from sklearn.preprocessing import OneHotEncoder
from sklearn import metrics

from typing import TypeVar
A = TypeVar('anndata')  ## generic for anndata
ENC = TypeVar('OneHotEncoder')

from TORC.models.MLP import MLP

# logger = logging.getLogger(__name__)

MLP_DIMS = [64, 16]
BATCH_SIZE = 32
Celltype_COLUMN = "cell.type"
PredCelltype_COLUMN = "pred_celltype"
ENTROPY_QUANTILE = 0.4  ## how many cells are used as second-round target


def _COOmtx_data_loader(mtx_prefix: str) -> A:
    '''
    Load gene score matrix in COO format
    ---
    Input:
        - mtx_prefix: gene score matrix in COO format
    ---
    Output:
        - an anndata object
    '''
    adata = anndata.read_mtx(mtx_prefix+'.mtx.gz').T
    genes = pd.read_csv(mtx_prefix+'_genes.tsv', header=None, sep='\t')
    adata.var["genes"] = genes[0].values
    adata.var_names = adata.var["genes"]
    adata.var_names_make_unique(join="-")
    adata.var.index.name = None
    cells = pd.read_csv(mtx_prefix+'_barcodes.tsv', header=None, sep='\t')
    adata.obs["barcode"] = cells[0].values
    adata.obs_names = adata.obs['barcode']
    adata.obs_names_make_unique(join="-")
    adata.obs.index.name = None

    adata = adata[:, adata.var_names.notnull()]
    adata.var_names=[i.upper() for i in list(adata.var_names)]
    return adata


def _csv_data_loader(csv_input: str) -> A:
    '''
    Load gene score matrix in csv format
    ---
    Input:
        - csv_input: gene scre matrix in dense format with row as genes and cells as columns.
    ---
    Output:
        - an anndata object
    '''
    df = pd.read_csv(csv_input, index_col=0)
    obs = pd.DataFrame(data=df.columns, index=df.columns)
    obs.columns = ["barcode"]
    var = pd.DataFrame(data=df.index, index=df.index)
    var.columns = ['gene_symbols']
    adata = anndata.AnnData(X=df.T, obs=obs, var=var)
    adata.obs_names_make_unique(join="-")
    adata.var_names_make_unique(join="-")

    adata = adata[:, adata.var_names.notnull()]
    adata.var_names=[i.upper() for i in list(adata.var_names)]
    return adata

def _metadata_loader(metadata):
    '''Load metadata
    '''
    metadata = pd.read_csv(metadata, index_col=0, sep=',')
    return metadata


def _process_adata(adata, process_type='train', celltype_label='celltype'):
    '''Procedures for filtering single-cell gene scale data (can be gene expression, or gene scores)
       1. Filter nonsense genes;
       2. Normalize and log-transform the data;
       3. Remove cells with no labels;
    '''
    adata = adata[:, adata.var_names.notnull()]  ## remove NA var_names, some genes generated by ArchR gene scores will be NA
    adata.var_names=[i.upper() for i in list(adata.var_names)] #avoid some genes having lower letter

    ## make names unique after removing
    adata.var_names_make_unique()
    adata.obs_names_make_unique()

    #prefilter_specialgene: MT and ERCC  -> refered from ItClust package
    Gene1Pattern="ERCC"
    Gene2Pattern="MT-"
    id_tmp1=np.asarray([not str(name).startswith(Gene1Pattern) for name in adata.var_names],dtype=bool)
    id_tmp2=np.asarray([not str(name).startswith(Gene2Pattern) for name in adata.var_names],dtype=bool)
    id_tmp=np.logical_and(id_tmp1,id_tmp2)
    adata._inplace_subset_var(id_tmp)

    ## handel exception when there are not enough cells or genes after filtering
    if adata.shape[0] < 3 or adata.shape[1] < 3:
        sys.exit("Error: too few genes or cells left to continue..")

    ## normalization,var.genes,log1p
    sc.pp.normalize_per_cell(adata, counts_per_cell_after=10000, min_counts=0)
    sc.pp.log1p(adata)

    ## cells with celltypes
    if process_type == 'train':
        cells = adata.obs.dropna(subset=[celltype_label]).index.tolist()
        adata = adata[cells]
    return adata

def _select_feature(adata: A, fs_method = "F-test", num_features: int = 1000) -> A:
    '''Select features
    ---
    Input:
        - anndata
        - fs_method: F-test / noFS / seurat
    '''
    ## Feature selection
    if fs_method == "noFS":
        print("TORC will not perform feature selection.\n")
        return adata
    else:
        if num_features > adata.shape[1]:
            print("Number of features is larger than data. TORC will not perform feature selection.\n")
            return adata

    if fs_method == "F-test":
        print("Use F-test to select features.\n")
        if scipy.sparse.issparse(adata.X) or \
                isinstance(adata.X, pd.DataFrame):
            tmp_data = adata.X.toarray()
        else:
            tmp_data = adata.X

        ## calculate F-test
        cell_annots = adata.obs[Celltype_COLUMN].tolist()
        uniq_celltypes = set(cell_annots)
        array_list = []
        for celltype in uniq_celltypes:
            idx = np.where(np.array(cell_annots) == celltype)[0].tolist()
            array_list.append(tmp_data[idx, :])
        F, p = scipy.stats.f_oneway(*array_list)
        F_updated = np.nan_to_num(F)
        sorted_idx = np.argsort(F_updated)[-num_features:]
        features = adata.var_names[sorted_idx].tolist()
        features.sort()
        adata = adata[:, features]

    if fs_method == "seurat":
        print("Use seurat in scanpy to select features.\n")
        sc.pp.highly_variable_genes(adata, n_top_genes=num_features, subset=True)
    return adata



def _scale_data(adata):
    '''Center scale
    '''
    sc.pp.scale(adata, zero_center=True, max_value=6)
    return adata

def _visualize_data(adata, output_dir, color_columns=["celltype"],
        reduction="tSNE", prefix="data"):
    '''Visualize data

    ---
    Input:
        - reduction: tSNE or UMAP
        - color_columns: plot on categories
    '''
    sc.tl.pca(adata, random_state=RANDOM_SEED)

    if reduction == "tSNE":
        sc.tl.tsne(adata, use_rep="X_pca",
            learning_rate=300, perplexity=30, n_jobs=1, random_state=RANDOM_SEED)
        sc.pl.tsne(adata, color=color_columns)
        plt.tight_layout()
        plt.savefig(output_dir+os.sep+prefix+"tSNE_cluster.png")
    if reduction == "UMAP":
        sc.pp.neighbors(adata, n_neighbors=20, use_rep="X_pca", random_state=RANDOM_SEED)
        sc.tl.umap(adata, random_state=RANDOM_SEED)
        sc.pl.umap(adata, color=color_columns)
        plt.tight_layout()
        plt.savefig(output_dir+os.sep+prefix+"umap_cluster.png")

def _save_adata(adata, output_dir, prefix=""):
    '''Save anndata as h5ad
    '''
    adata.write(output_dir+os.sep+prefix+'adata.h5ad')


def _prob_to_label(y_pred: np.ndarray, encoders: dict) -> list:
    '''Turn predicted probabilites to labels
    ---
    Input:
        - y_pred: Predicted probabilities
        - encoders: dictionary with mapping information
    ---
    Output:
        - a list containing predicted cell types
    '''
    pred_labels = y_pred.argmax(1)
    pred_celltypes = [encoders[label] for label in pred_labels]
    print("=== Predicted celltypes: ", set(pred_celltypes))
    return pred_celltypes

def _label_to_onehot(labels: list, encoders:dict) -> np.ndarray:
    '''Turn predicted labels to onehot encoder
    ---
    Input:
        - labels: the input predicted cell types
        - encoders: dictionary with mapping information
    '''
    inv_enc = {v: k for k, v in encoders.items()}
    onehot_arr = np.zeros((len(labels), len(encoders)))
    pred_idx = [inv_enc[l] for l in labels]
    onehot_arr[np.arange(len(labels)), pred_idx] = 1
    return onehot_arr


def _extract_adata(adata: A) -> np.ndarray:
    '''Extract adata.X to a numpy array
    ---
    Output:
         - matrix in np.ndarray format
    '''
    if scipy.sparse.issparse(adata.X) or isinstance(adata.X, pd.DataFrame) or isinstance(adata.X, anndata._core.views.ArrayView):
        X = adata.X.toarray()
    else:
        X = adata.X
    return X

def _init_MLP(x_train, y_train, dims=[64, 16], seed=0):
    '''Initialize MLP model based on input data
    '''
    mlp = MLP(dims)
    mlp.input_shape = (x_train.shape[1], )
    #mlp.n_classes = len(set(y_train.argmax(1)))
    mlp.n_classes = y_train.shape[1]
    mlp.random_state = seed
    mlp.init_MLP_model()  ## init the model
    return mlp

def _select_confident_cells(adata, celltype_col):
    '''Select low entropy cells from each predicted cell type
    ---
    Input:
        - adata: anndata object
        - celltype_col: the column indicator
    '''
    low_entropy_cells = []
    for celltype in set(adata.obs[celltype_col]):
        celltype_df = adata.obs[adata.obs[celltype_col] == celltype]
        entropy_cutoff = np.quantile(celltype_df['entropy'], q=ENTROPY_QUANTILE)
        ## change to < instead of <= to deal with ties
        cells = celltype_df.index[np.where(celltype_df['entropy'] <= entropy_cutoff)[0]].tolist()
        num_cells = math.ceil(ENTROPY_QUANTILE*celltype_df.shape[0])
        if len(cells) > num_cells:
            random.seed(RANDOM_SEED)
            selected_cells = random.sample(cells, num_cells)
        else:
            selected_cells = cells
        low_entropy_cells.extend(selected_cells)
    high_entropy_cells = list(set(adata.obs_names) - set(low_entropy_cells))
    adata.obs.loc[low_entropy_cells, 'entropy_status'] = "low"
    adata.obs.loc[high_entropy_cells, 'entropy_status'] = "high"
    return adata

def _run_pipeline(train_adata, test_adata, result_dir, rseed=1234):

    RANDOM_SEED = rseed
    random.seed(RANDOM_SEED)
    np.random.seed(RANDOM_SEED)

    ## Train model
    x_train = _extract_adata(train_adata)
    enc = OneHotEncoder(handle_unknown='ignore')
    y_train = enc.fit_transform(train_adata.obs[["cell.type"]]).toarray()
    logger.debug("Categories information: ", enc.categories_[0])

    mlp = _init_MLP(x_train, y_train, dims=MLP_DIMS,
            seed=RANDOM_SEED)
    mlp.compile()
    mlp.fit(x_train, y_train)

    with open(result_dir+os.sep+"onehot_encoder.txt", 'w') as f:
        for idx, cat in enumerate(enc.categories_[0]):
            f.write('%d:%s\n' % (idx, cat))

    model = mlp.model
    encoders = {}
    with open(result_dir+os.sep+"onehot_encoder.txt") as f:
        for line in f:
            line_info = line.strip().split(':')
            encoders[int(line_info[0])] = line_info[1]

    ## Check fit
    y_fit = tf.nn.softmax(model.predict(x_train)).numpy()
    fit_celltypes = _prob_to_label(y_fit, encoders)
    fit_COLUMN = 'fit_' + 'celltype'
    train_adata.obs[fit_COLUMN] = fit_celltypes

    ## select certain columns and store to the file
    train_adata.obs.to_csv(result_dir+os.sep+"train_obs.csv")


    ## Naive MLP
    test_data_mat = _extract_adata(test_adata)
    y_pred = tf.nn.softmax(model.predict(test_data_mat)).numpy()
    pred_celltypes = _prob_to_label(y_pred, encoders)
    test_adata.obs[PredCelltype_COLUMN] = pred_celltypes

    y_avg_pred = np.average(y_pred,axis=0)
    df = pd.DataFrame(y_avg_pred)
    df.columns = ["prob"]
    df.index = enc.categories_[0]
    df.to_csv(result_dir+os.sep+'avg_post_prob.csv')

    firstround_COLUMN = 'firstround_' + 'pred_celltype'
    test_adata.obs[firstround_COLUMN] = pred_celltypes



    entropy = [-np.nansum(y_pred[i]*np.log(y_pred[i])) for i in range(y_pred.shape[0])]
    test_adata.obs['entropy'] = entropy
    test_adata = _select_confident_cells(
                test_adata, celltype_col=firstround_COLUMN)

    ## Select cells based on Entropy
    low_entropy_cells = test_adata.obs_names[np.where(test_adata.obs['entropy_status'] == 'low')].tolist()
    high_entropy_cells = test_adata.obs_names[np.where(test_adata.obs['entropy_status'] == 'high')].tolist()
    test_ref_adata = test_adata[low_entropy_cells]
    test_tgt_adata = test_adata[high_entropy_cells]

    x_tgt_train = _extract_adata(test_ref_adata)
    y_tgt_train = _label_to_onehot(test_ref_adata.obs.loc[low_entropy_cells, firstround_COLUMN].tolist(),
                    encoders=encoders)
    x_tgt_test = _extract_adata(test_tgt_adata)

    ## Second MLP
    teacher = _init_MLP(x_tgt_train, y_tgt_train, dims=Teacher_DIMS,
                seed=RANDOM_SEED)
    teacher.compile()
    teacher.fit(x_tgt_train, y_tgt_train, batch_size=BATCH_SIZE)
    ## student model -> actually same model, just used the concept of distillation
    student = _init_MLP(x_tgt_train, y_tgt_train, dims=Student_DIMS,
                    seed=RANDOM_SEED)
    # Initialize and compile distiller
    distiller = _run_distiller(x_tgt_train, y_tgt_train,
                    student_model=student.model,
                    teacher_model=teacher.model)

    ## Check fit
    y_fit_tgt = tf.nn.softmax(model.predict(x_tgt_train)).numpy()
    fit_celltypes_tgt = _prob_to_label(y_fit_tgt, encoders)
    fit_COLUMN = 'fit_' + 'celltype'
    test_adata.obs[fit_COLUMN] = pred_celltypes
    test_adata.obs.loc[low_entropy_cells, fit_COLUMN] = fit_celltypes_tgt


    y_pred_tgt = tf.nn.softmax(distiller.student.predict(x_tgt_test)).numpy()

    pred_celltypes = _prob_to_label(y_pred_tgt, encoders)
    test_adata.obs.loc[high_entropy_cells, PredCelltype_COLUMN] = pred_celltypes

    ## select certain columns and store to the file
    test_adata.obs.to_csv(result_dir+os.sep+"predicted_obs.csv")

    acc = metrics.accuracy_score(test_adata.obs["cell.type"], test_adata.obs["firstround_pred_celltype"])
    ARI = metrics.cluster.adjusted_rand_score(test_adata.obs["cell.type"], test_adata.obs["firstround_pred_celltype"])
    macroF1 = metrics.f1_score(test_adata.obs["cell.type"], test_adata.obs["firstround_pred_celltype"], average="macro")
    print("*****=====", "Testing accuracy: ", acc, " Testing ARI: ", ARI,
            "Testing MacroF1: ", macroF1, "=====*****")

    with open(result_dir+os.sep+"metrics.txt", 'w') as f:
        f.write("Acc:%s\n" % str(acc))
        f.write("ARI:%s\n" % str(ARI))
        f.write("macroF1:%s\n" % str(macroF1))


    acc = metrics.accuracy_score(test_adata.obs["cell.type"], test_adata.obs["pred_celltype"])
    ARI = metrics.cluster.adjusted_rand_score(test_adata.obs["cell.type"], test_adata.obs["pred_celltype"])
    macroF1 = metrics.f1_score(test_adata.obs["cell.type"], test_adata.obs["pred_celltype"], average="macro")
    print("*****=====", "Testing accuracy: ", acc, " Testing ARI: ", ARI,
            "Testing MacroF1: ", macroF1, "=====*****")

    with open(result_dir+os.sep+"Second_metrics.txt", 'w') as f:
        f.write("Acc:%s\n" % str(acc))
        f.write("ARI:%s\n" % str(ARI))
        f.write("macroF1:%s\n" % str(macroF1))
